core/src/main/java/com/microsoft/applicationinsights/internal/heartbeat/HeartBeatProvider.java

```diff
@@ -0,0 +1,298 @@
+    this.heartbeatProperties = new ConcurrentHashMap<>();
+    this.isEnabled = true;
+    this.heartbeatsSent = new AtomicLong(0);
+  }
+
+  @Override
+  public String getInstrumentationKey() {
+    return this.telemetryClient.getContext().getInstrumentationKey();
+  }
+
+  @Override
+  public void setInstrumentationKey(String key) {
+    if (this.telemetryClient != null && key != null) {
+      this.telemetryClient.getContext().setInstrumentationKey(key);
+    }
+  }
+
+  @Override
+  public void initialize(TelemetryConfiguration configuration) {
+    if (isEnabled) {
+      if (this.telemetryClient == null) {
+        this.telemetryClient = new TelemetryClient(configuration);
+      }
+
+      //Submit task to set properties to dictionary using separate thread. we do not wait for the
+      //results to come out as some I/O bound properties may take time.
+      executorService.submit(HeartbeatDefaultPayload.populateDefaultPayload(getExcludedHeartBeatProperties(),
+          getExcludedHeartBeatPropertyProviders(), this));
+
+
+
+      scheduledExecutorService.scheduleAtFixedRate(heartBeatPulse(), interval, interval, TimeUnit.SECONDS);
+    }
+  }
+
+  @Override
+  public boolean addHeartBeatProperty(String propertyName, String propertyValue,
+      boolean isHealthy) {
+
+    boolean isAdded= false;
+    if (!StringUtils.isEmpty(propertyName)) {
+      try {
+        if (!heartbeatProperties.containsKey(propertyName)) {
+             HeartBeatPropertyPayload payload = new HeartBeatPropertyPayload();
+             payload.setHealthy(isHealthy);
+             payload.setPayloadValue(propertyValue);
+             heartbeatProperties.put(propertyName, payload);
+             isAdded = true;
+             InternalLogger.INSTANCE.trace("added heartbeat property");
+        }
+        else {
+          throw new Exception("heartbeat property cannot be added twice");
+        }
+      } catch (Exception e) {
+        InternalLogger.INSTANCE.warn("Failed to add the property %s value %s, stack trace is : %s," ,
+            propertyName, propertyValue, ExceptionUtils.getStackTrace(e));
+      }
+    }
+    else {
+      InternalLogger.INSTANCE.warn("cannot add property without property name");
+    }
+    return isAdded;
+  }
+
+  @Override
+  public boolean setHeartBeatProperty(String propertyName, String propertyValue,
+      boolean isHealthy) {
+
+    boolean setResult = false;
+    if (!StringUtils.isEmpty(propertyName)) {
+      try {
+        if (heartbeatProperties.containsKey(propertyName) && !HeartbeatDefaultPayload.isDefaultKeyWord(propertyName)) {
+          HeartBeatPropertyPayload payload = new HeartBeatPropertyPayload();
+          payload.setHealthy(isHealthy);
+          payload.setPayloadValue(propertyValue);
+          heartbeatProperties.put(propertyName, payload);
+          setResult = true;
+        }
+        else {
+          throw new Exception("heartbeat property cannot be set without adding it first, or heartbeat"
```

Comment:
This exception will be thrown if 1. Property is not added first. 2. propertyname is a keyword.
Would be ideal to separate